src/olmo_core/nn/conversion/state_mapping.py

```diff
@@ -0,0 +1,250 @@
+from dataclasses import dataclass
+from typing import Any, Dict, Optional, Set, Tuple
+
+from olmo_core.config import StrEnum
+
+
+class TemplatePlaceholder(StrEnum):
+    LAYER = "[layer]"
+    EXPERT = "[expert]"
+
+
+@dataclass
+class StateMappingTemplate:
+    """
+    The template for a mapping from state from one format to another format (e.g. OLMo Core to HF).
+    These mappings are 'templates' since they support keys and other metadata having placeholders
+    for information like the layer number or number of MoE experts. This class can be converted
+    to a `StateMapping` by providing the placeholder information information.
```

Comment:
minor typos:
```suggestion
    The template for a mapping state from one format to another format (e.g. OLMo Core to HF).
    These mappings are 'templates' since they support keys and other metadata having placeholders
    for information like the layer number or number of MoE experts. This class can be converted
    to a `StateMapping` by providing the placeholder information.
```